File: modules/wiesenjournal/backend/app/nutrients.py
"""Nährstoffrechnung je Düngungsmassnahme — Zwilling von
frontend/src/lib/nutrients.ts (keep in sync!). Menge × Gehalt je Einheit
der Düngerart; bei Gülle skaliert der Verdünnungsfaktor (Gülle-Anteil im
ausgebrachten Volumen) relativ zum Default der Düngerart: RGv-Werte gelten
für 1:1 (0.5) — wird 3:1 (0.75) ausgebracht, ist pro m³ 1.5× so viel drin.
"""

from __future__ import annotations

from decimal import Decimal
from typing import Any


def parse_dilution(text: str | None) -> float | None:
    """'3:1' (Gülle:Wasser) → 0.75; '1:1' → 0.5; None/unlesbar → None."""
    if not text:
        return None
    parts = text.replace(" ", "").split(":")
    if len(parts) != 2:
        return None
    try:
        a, b = float(parts[0]), float(parts[1])
    except ValueError:
        return None
    return a / (a + b) if a + b > 0 else None


def _f(v: Any) -> float:
    if v is None:
        return 0.0
    if isinstance(v, Decimal):
        return float(v)
    return float(v)


def compute_nutrients(amount: float | None, ftype: dict, dilution_factor: float | None) -> dict[str, float | None]:
    if amount is None:
        return {"n_kg": None, "n_avail_kg": None, "p2o5_kg": None, "k2o_kg": None}
    default = _f(ftype.get("dilution_default")) or 1.0
    scale = (dilution_factor / default) if dilution_factor else 1.0
    eff = amount * scale
    n_kg = eff * _f(ftype.get("n_kg_per_unit"))
    return {
        "n_kg": round(n_kg, 2),
        "n_avail_kg": round(n_kg * _f(ftype.get("n_avail_pct")) / 100.0, 2),
        "p2o5_kg": round(eff * _f(ftype.get("p2o5_kg_per_unit")), 2),
        "k2o_kg": round(eff * _f(ftype.get("k2o_kg_per_unit")), 2),
    }

File: modules/wiesenjournal/frontend/src/lib/nutrients.ts
import { num } from './format'
import type { FertilizerType } from '../types'

// Zwilling von backend/app/nutrients.py — keep in sync!
// Menge × Gehalt je Einheit der Düngerart; bei Gülle skaliert der
// Verdünnungsfaktor (Gülle-Anteil im ausgebrachten Volumen) relativ zum
// Default der Düngerart: RGv-Werte gelten für 1:1 (0.5) — wird 3:1 (0.75)
// ausgebracht, ist pro m³ 1.5× so viel drin.

export interface Nutrients {
  n_kg: number | null
  n_avail_kg: number | null
  p2o5_kg: number | null
  k2o_kg: number | null
}

/** '3:1' (Gülle:Wasser) → 0.75; '1:1' → 0.5; leer/unlesbar → null. */
export function parseDilution(text: string | null | undefined): number | null {
  if (!text) return null
  const parts = text.replace(/\s/g, '').split(':')
  if (parts.length !== 2) return null
  const a = Number(parts[0])
  const b = Number(parts[1])
  if (Number.isNaN(a) || Number.isNaN(b) || a + b <= 0) return null
  return a / (a + b)
}

function round2(v: number): number {
  return Math.round(v * 100) / 100
}

export function computeNutrients(
  amount: number | null,
  type: FertilizerType | null,
  dilutionFactor: number | null,
): Nutrients {
  if (amount == null || !type) return { n_kg: null, n_avail_kg: null, p2o5_kg: null, k2o_kg: null }
  const dflt = num(type.dilution_default) || 1
  const scale = dilutionFactor ? dilutionFactor / dflt : 1
  const eff = amount * scale
  const nKg = eff * (num(type.n_kg_per_unit) ?? 0)
  return {
    n_kg: round2(nKg),
    n_avail_kg: round2((nKg * (num(type.n_avail_pct) ?? 0)) / 100),
    p2o5_kg: round2(eff * (num(type.p2o5_kg_per_unit) ?? 0)),
    k2o_kg: round2(eff * (num(type.k2o_kg_per_unit) ?? 0)),
  }
}

/** kg/ha aus kg und Aren (100 a = 1 ha). */
export function kgPerHa(kg: number | null | undefined, areaA: number | null | undefined): number | null {
  if (kg == null || !areaA) return null
  return round2((kg / areaA) * 100)
}
